### DISCUSSION TITLE
How to do mocks when testing a nodejs custom environment to test a fastify server?

### DISCUSSION THREAD

**Discussion by akhilmhdh** [vitest-dev/vitest#6365] [Q&A] [answered]
Hey everyone,

We are using vitest to test our fastify nodejs server. How we do is that is using the vitest custom environment option.

So our custom environment file would looks like this

```ts
// load environment
export default {
  name: "env",
  transformMode: "ssr",
  async setup() {
    // connect with db

    try {
      // do migration
     // do seeding

      const server = await main({ db, smtp, logger, queue, keyStore });
      // @ts-expect-error type
      globalThis.testServer = server;
    } catch (error) {
      console.log("[TEST] Error setting up environment", error);
      await db.destroy();
      throw error;
    }

    console.log("ended environment");
    // custom setup
    return {
      async teardown() {
        await globalThis.testServer.close();
        await db.destroy();
      }
    };
  }
};
```

A simple spec file would look like this
```
describe("Org V1 Router", async () => {
  test("GET Org list", async () => {
    const res = await testServer.inject({
      method: "GET",
      url: "",
      headers: {
        authorization: `Bearer ${<valid token>}`
      }
    });
    expect(res.statusCode).toBe(200);
    const payload = JSON.parse(res.payload);
    expect(payload).toHaveProperty("organizations");
  });
});
```

This seems to be working really well for us. Until we wanted some mocking to do on testing on endpoints. Like axios as an example.

For some reason none of mocks are working. Like ones setup in `setupFiles` nor anywhere in spec? I do think i am making a mistake in doing mocking (as i am mainly amateur in test frameworks and my experience in mocking is not that great 😅 ) or is mocking not possible with this way?

Any insights would be helpful!

**Comment by hi-ogawa** [comment, marked as answer]
I think environment loading happens before any module graph setup (namely `VitestExecutor`), so probably that's why `vi.mock` wouldn't work.

Probably you had some reasons to choose custom environment approach, but just in case, sharing another way to achieve similar server setup using higher level API (so it should compose better with other APIs like `vi.mock`).
Have you considered something like this using `beforeAll` in `setupFiles`?

```ts
beforeAll(async () => {
  // connect with db

  try {
    // do migration
    // do seeding

    const server = await main({ db, smtp, logger, queue, keyStore });
    globalThis.testServer = server;
  } catch (error) {
    console.log("[TEST] Error setting up environment", error);
    await db.destroy();
    throw error;
  }

  console.log("ended environment");
  // custom setup
  return async () => {
    await globalThis.testServer.close();
    await db.destroy();
  }
})
```

**Reply by akhilmhdh** [reply]
Yes it seems your right. The environment as I guessed doesn't allow mocking to work due to module loading. I guess that's why `.resetModule` is also not working.

We went for custom environment because we wanted to do the process only once and not for each test suite. Something like the `globalSetup`. But globalSetup doesn't allow the `globalThis` because i think its run on a different context and the `inject` from vitest only allows JSON serializable.

**Reply by hi-ogawa** [reply]
> We went for custom environment because we wanted to do the process only once and not for each test suite.

By default, environment setup is isolated for each test file, so I was assuming that `beforeAll` would mostly work same, but perhaps you disable isolation by `isolate: false`? Sharing your config would help to understand your use case better.

**Reply by akhilmhdh** [reply]
Sure,
Here you go: [vitest config](https://github.com/Infisical/infisical/blob/main/backend/vitest.e2e.config.ts)

Yes we have single thread and isolate as false. Because the issue is migration as multiple entity cannot run it at the same time. 

We thought of making it as start of the script but the issue is cleanup causes conflict too.

**Reply by hi-ogawa** [reply]
Thanks for sharing the full repo, that's even better :+1: 

It might be too simplistic, but maybe you can still delay db/server setup till `setupFiles + beforeAll` by checking `globalThis.testServer`?

```ts
beforeAll(async () => {
  if (globalThis.testSever) return;

  // then setup db / server only on first setup
});
```

**Reply by akhilmhdh** [reply]
Yes I need to rethink our implementation once again but i think i now have a better understanding of vitest. 

I did try the setupFiles way and the mocks were indeed working. So marking above as the answer indeed. Thank you so much for your valuable time and insight. Have a good day.
